### PR TITLE
Simplify optimized allocation

### PR DESCRIPTION
**Pull request by jacob-hughes**
We can now use the `std::gc::needs_tracing` and
`std::gc::can_trace_precisely` intrinsics to make allocation decisions.
These intrinsics are const expressions, so have no branching cost. This
lets us remove the complicated specialization stuff.

### CODE REVIEW

File: src/allocator.rs [hunk 1]
```diff
@@ -33,39 +28,89 @@ unsafe impl GlobalAlloc for GcAllocator {
...
 }
 
 impl GcAllocator {
-    #[cfg(feature = "rustgc_internal")]
+    /// Allocate `T` such that it is optimized for marking.
+    #[cfg(feature = "rustgc")]
     pub fn maybe_optimised_alloc<T>(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
-        let sp = specializer::AllocationSpecializer::new();
-        sp.maybe_optimised_alloc::<T>(layout)
+        // assert_eq!(Layout::new::<T>, layout);
```
Comment (ltratt) [inline src/allocator.rs]: Commented code.

**Comment by ltratt** [comment]
Looks OK apart from that one small comment!

File: src/allocator.rs (hunk 1 above)
Comment (jacob-hughes) [reply src/allocator.rs]: That's embarrassing!

**Comment by jacob-hughes** [comment]
Fixed: 422911c61a3d7521b04534777d11dadd8648f7eb

**Comment by ltratt** [comment]
I think you'll need to `cargo fmt` away the resulting blank line?

**Comment by jacob-hughes** [comment]
The blank line was intentional, as I wanted the assertion to stand on it's own. It passes the fmt check, but I can remove it if you'd prefer?

**Comment by ltratt** [comment]
Doh, you uncommented it -- I thought you'd deleted it! I'm an idiot. Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by jacob-hughes** [comment]
This one failed because it needed the other one to be merged first. Can you try again ?

**Comment by jacob-hughes** [comment]
This build failure has inadvertently found the bug I was scratching my head for days over with the Rust codegen tests not optimizing away alloc calls. 

It turns out, there is an invariant in LLVM that these `MemoryBuiltins` need to be defined in source order, but I never picked up on this because only `libgc` builds and compiles `rustgc` with LLVM assertions enabled.

**Comment by ltratt** [comment]
Is it fixable within this PR?

**Comment by jacob-hughes** [comment]
No, we'll need to land https://github.com/softdevteam/rustgc-llvm-fork/pull/3 (and then a submodule bump in rustgc) first.

**Comment by ltratt** [comment]
Sorry, I just saw that we need a submodule bump somewhere.

**Comment by jacob-hughes** [comment]
Ok this one should merge now :crossed_fingers:

**Pull request merged**